sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp

```diff
@@ -0,0 +1,777 @@
+            Device->ZeDeviceComputeProperties->maxGroupSizeZ};
+        GroupSize[I] = std::min(size_t(GroupSize[I]), GlobalWorkSize[I]);
+        while (GlobalWorkSize[I] % GroupSize[I]) {
+          --GroupSize[I];
+        }
+        if (GlobalWorkSize[I] / GroupSize[I] > UINT32_MAX) {
+          urPrint("urCommandBufferAppendKernelLaunchExp: can't find a WG size "
+                  "suitable for global work size > UINT32_MAX\n");
+          return UR_RESULT_ERROR_INVALID_WORK_GROUP_SIZE;
+        }
+        WG[I] = GroupSize[I];
+      }
+      urPrint(
+          "urCommandBufferAppendKernelLaunchExp: using computed WG size = {%d, "
+          "%d, %d}\n",
+          WG[0], WG[1], WG[2]);
+    }
+  }
+
+  // TODO: assert if sizes do not fit into 32-bit?
+  switch (WorkDim) {
+  case 3:
+    ZeThreadGroupDimensions.groupCountX =
+        ur_cast<uint32_t>(GlobalWorkSize[0] / WG[0]);
+    ZeThreadGroupDimensions.groupCountY =
+        ur_cast<uint32_t>(GlobalWorkSize[1] / WG[1]);
+    ZeThreadGroupDimensions.groupCountZ =
+        ur_cast<uint32_t>(GlobalWorkSize[2] / WG[2]);
+    break;
+  case 2:
+    ZeThreadGroupDimensions.groupCountX =
+        ur_cast<uint32_t>(GlobalWorkSize[0] / WG[0]);
+    ZeThreadGroupDimensions.groupCountY =
+        ur_cast<uint32_t>(GlobalWorkSize[1] / WG[1]);
+    WG[2] = 1;
+    break;
+  case 1:
+    ZeThreadGroupDimensions.groupCountX =
+        ur_cast<uint32_t>(GlobalWorkSize[0] / WG[0]);
+    WG[1] = WG[2] = 1;
+    break;
+
+  default:
+    urPrint("urCommandBufferAppendKernelLaunchExp: unsupported work_dim\n");
+    return UR_RESULT_ERROR_INVALID_VALUE;
+  }
+
+  // Error handling for non-uniform group size case
+  if (GlobalWorkSize[0] !=
+      size_t(ZeThreadGroupDimensions.groupCountX) * WG[0]) {
+    urPrint("urCommandBufferAppendKernelLaunchExp: invalid work_dim. The range "
+            "is not a "
+            "multiple of the group size in the 1st dimension\n");
+    return UR_RESULT_ERROR_INVALID_WORK_GROUP_SIZE;
+  }
+  if (GlobalWorkSize[1] !=
+      size_t(ZeThreadGroupDimensions.groupCountY) * WG[1]) {
+    urPrint("urCommandBufferAppendKernelLaunchExp: invalid work_dim. The range "
+            "is not a "
+            "multiple of the group size in the 2nd dimension\n");
+    return UR_RESULT_ERROR_INVALID_WORK_GROUP_SIZE;
+  }
+  if (GlobalWorkSize[2] !=
+      size_t(ZeThreadGroupDimensions.groupCountZ) * WG[2]) {
+    urPrint("urCommandBufferAppendKernelLaunchExp: invalid work_dim. The range "
+            "is not a "
+            "multiple of the group size in the 3rd dimension\n");
+    return UR_RESULT_ERROR_INVALID_WORK_GROUP_SIZE;
+  }
+
+  return UR_RESULT_SUCCESS;
+}
+
+/// Helper function to take a list of ur_exp_command_buffer_sync_point_ts and
+/// fill the provided vector with the associated ZeEvents
+static ur_result_t getEventsFromSyncPoints(
+    const std::unordered_map<ur_exp_command_buffer_sync_point_t,
+                             ur_event_handle_t> &SyncPoints,
```

Comment:
what is the difference between SyncPointrs and SyncPointWaitList?

Reply:
`SyncPoints` is all the sync-points in a command-buffer, and their mapping to UR events. While `SyncPointWaitList` is a list of sync-points that the dependencies of a new command, and a subset of `SyncPoints` to lookup. 

I commented and tweaked this function to make it clearer, so that we pass the command-buffer itself for the first argument rather than two sync-point containers
https://github.com/intel/llvm/blob/cf3a2bd79964b2dfc7fae141f6eae78fd73679e9/sycl/plugins/unified_runtime/ur/adapters/level_zero/ur_level_zero_command_buffer.cpp#L234-L265